### DISCUSSION TITLE
Is there a way to manually invalidate useLoaderData?

### DISCUSSION THREAD

**Discussion by guillaumebrunerie** [remix-run/remix#3940] [Q&A] [answered]
I'm making a website (say a todo list) where if an item is checked/added/removed by a client, then every other client connected to the same todo list should see the changes immediately, without having to do anything.
I have set up WebSockets and now I want the client to invalidate (refetch) the data returned by `useLoaderData` as soon as it receives a message on the WebSocket.
Is there a way to do that?
I managed to hack my way towards it by combining `useLoaderData` for the initial load and `useFetcher` for WebSocket-initiated reloads, see incomplete code below.

```js
export default function ListPage() {
	const fetcher = useFetcher<LoaderData>();
	const loaderData = useLoaderData<LoaderData>();
	const data = fetcher.data || loaderData;

	const socket = useSocket();
	const fetcherLoad = fetcher.load;
	React.useEffect(() => {
		if (!socket) return;

		socket.on("update", (updatedListId) => {
			if (updatedListId === listId) {
				fetcherLoad(`/${listId}?index`);
			}
		});
	}, [socket, fetcherLoad, listId]);

	// Do stuff with `data`
}
```

It works for the use case described above, but it breaks when navigating to another list because now only the data from `useLoaderData` is updated, so if there has been a `fetcher.load` in the mean time, my code won’t see the change to `loaderData`.
I guess I could write more code to fix this problem, but it feels like I’m reinventing the wheel, surely there is a simpler way to do it?

**Comment by guillaumebrunerie** [comment]
Found another partial solution, using `useNavigate` and navigating to the same page we are currently on, it works properly when it comes to reloading the data, but it scrolls back up to the top of the page every time, which is kinda annoying.

**Comment by corygibbons** [comment, marked as answer]
This may load more data than you want but previously in a similar case I used an empty action since I could count on `useLoaderData` to always be accurate, plus it'll preserve the scroll location. Something like this (untested):

```js
export const action = async () => {
  return json({})
}

export default function ListPage() {
	const fetcher = useFetcher<LoaderData>();
	const loaderData = useLoaderData<LoaderData>();
	const data = fetcher.data || loaderData;

	const socket = useSocket();

	React.useEffect(() => {
		if (!socket) return;

		socket.on("update", (updatedListId) => {
			if (updatedListId === listId) {
				fetcher.submit({}, { method: 'post' });
			}
		});
	}, [socket, fetcherLoad, listId]);

	// Do stuff with `data`
}
```

**Reply by guillaumebrunerie** [reply]
Thank you, works great!
